src/extractor/guidance/sliproad_handler.cpp

```diff
@@ -29,273 +32,720 @@ SliproadHandler::SliproadHandler(const IntersectionGenerator &intersection_gener
+        auto onto_roundabout = hasRoundaboutType(road.instruction);
 
-    const auto source_edge_data = node_based_graph.GetEdgeData(source_edge_id);
-    // check whether the continue road is valid
-    const auto check_valid = [this, source_edge_data](const ConnectedRoad &road) {
-        const auto road_edge_data = node_based_graph.GetEdgeData(road.eid);
-        // Test to see if the source edge and the one we're looking at are the same road
+        if (onto_roundabout)
+            return false;
 
-        const auto same_name = !util::guidance::requiresNameAnnounced(
-            road_edge_data.name_id, source_edge_data.name_id, name_table, street_name_suffix_table);
+        const auto road_data = node_based_graph.GetEdgeData(road.eid);
 
-        return road_edge_data.road_classification == source_edge_data.road_classification &&
-               road_edge_data.name_id != EMPTY_NAMEID && source_edge_data.name_id != EMPTY_NAMEID &&
-               same_name && road.entry_allowed;
+        auto is_roundabout = road_data.roundabout;
+
+        if (is_roundabout)
+            return false;
+
+        return true;
     };
 
-    if (!check_valid(next_road))
+    if (!std::any_of(begin(intersection), end(intersection), is_potential_link))
         return intersection;
 
-    // Threshold check, if the intersection is too far away, don't bother continuing
-    const auto coordinate_extractor = intersection_generator.GetCoordinateExtractor();
-    const auto next_road_length = util::coordinate_calculation::getLength(
-        coordinate_extractor.GetForwardCoordinatesAlongRoad(
-            node_based_graph.GetTarget(source_edge_id), next_road.eid),
-        &util::coordinate_calculation::haversineDistance);
-    if (next_road_length > MAX_SLIPROAD_THRESHOLD)
+    // If the intersection is too far away, don't bother continuing
+    if (nextIntersectionIsTooFarAway(intersection_node_id, next_road.eid))
     {
         return intersection;
     }
-    auto next_intersection_node = node_based_graph.GetTarget(next_road.eid);
 
-    const auto next_road_next_intersection =
-        findNextIntersectionForRoad(intersection_node_id, next_road, next_intersection_node);
+    // Try to find the intersection at (c) which the Sliproad shortcuts
+    const auto next = getNextIntersection(intersection_node_id, next_road);
 
-    if (next_road_next_intersection.empty())
+    if (!next)
         return intersection;
 
     // If we are at a traffic loop at the end of a road, don't consider it a sliproad
-    if (intersection_node_id == next_intersection_node)
+    if (intersection_node_id == next->node)
         return intersection;
 
-    std::unordered_set<NameID> target_road_names;
+    std::vector<NameID> target_road_name_ids;
+    target_road_name_ids.reserve(next->intersection.size());
 
-    for (const auto &road : next_road_next_intersection)
+    for (const auto &road : next->intersection)
```

Comment:
can we potentially change `next` into `path_along_main_road` or something similar?